### DISCUSSION TITLE
DAO algorithm to calculate votes

### DISCUSSION THREAD

**Discussion by geopor** [tearust/teaproject#130] [Ideas]
## DAO Voting Design Philosophy
- The more a wallet address stakes their TEA into the platform for the benefit / functioning of the ecosystem, the more they'll be rewarded with voting points (which we call reputation tokens).
- We'll calculate voting power based on amount of TEA staked and for how long.
- Free TEA that isn't being "staked" into the ecosystem represents non-commitment for that wallet address and won't be awarded any voting privileges. Similarly, endusers paying gas to use a hosting node / TApp payment to use TApp get the benefit of using that app and won't be awarded voting privileges.
- Voting power is kept in TEA's layer-2 and are referred to as "reputation tokens". These are non-transferrable / bound to an ETH address. Basically a value that we store for each address in our layer-2 database
- Reputation tokens are burned when a user votes, i.e. once a user votes on an issue, they're removed from their wallet.
- For any vote, a user can devote more of their reputation tokens towards voting on that referendum (yes or no) which would increase their conviction on the issue.
- There will be a multiplier determining how quickly one accumulates reputation tokens. If a user has 0 reputation tokens, they'll accumulate quicker than someone with 10 reputation tokens. Someone with 10 reputation tokens accumulates faster than someone with 20 tokens etc. There will be a max reputation token point where people can't acccumulate them any more. This will encourage people to vote.
- Users will propose votes / vote directly at snapshot.org without an intermediate forum / Discord discussion.
- Special one-time provision: At mainnet launch, COFFEE will too.

### 1. Miners (node runners) earn reputation tokens
There are two types of miners in the TEA Project and both can earn reputation tokens.

1. State maintainer node operators earn reputation tokens in proportion to:

- Their locked TEA deposit 
- The daily Harberger Tax they pay for the privilege of running their node

2. CML miners earn reputation tokens only if they pay the state subscription fee.

- These CML miners earn reputation tokens in proportion to the state subscription fee they pay.

### 2. TApp application developers earn reputation tokens
A TApp owner will earn reputation tokens in proportion to how much memory tax they pay. Devs of TApps that don't pay any memory tax won't get any reputation tokens from their TApps.

### 3. Investors (bonding curve token holders) earn reputation tokens
Investors who stake their TEA into the bonding curve tokens are rewarded with reputation tokens.

- TApp tokens
- CML tokens

## Weighting the allocation of reputation tokens by time held
All of the above user classes will receive reputation tokens, but we have to prevent gaming the system (e.g. buying tokens just to get reputation tokens and then selling soon after count is taken).

Proposed solution:

- Randomly check balance during the day(?). If not day, the duration slot will be fixed but the check time will be random.

Also as mentioned earlier, those with less reputation tokens will accumulate them at a quicker rate than someone with more reputation tokens all things being equal.

We can use the following formula to calculate the reputation token gain per day:

```
( invested_tea_in_all_bonding_curve / 365 
 + node_owner_paid_state_sub_tax
iu + state_maintainer_paid_harberger_tax
 + app_owner_paid_memory_tax )
* square_root(your_reputation_balance)/your_reputation_balance
```

This formula implies that the higher a user's reputation balance, the slower they'll gain any new reputation. 

**One rule regarding accrual of reputation tokens:** If a user hasn't either voted on or proposed any DAO voting referendums for X months, their reputation will stop growing. 

## Voting Algorithm
We'll use the Polkadot GOV2 algorithm for voting:

<img width="404" alt="Untitled 2 11 41 49 AM" src="https://user-images.githubusercontent.com/86096370/206794032-0c230459-af2c-4443-8b40-07d43828f65f.png">

## Technical Fellowship
Not everything is up for vote, notably the technical decisions that govern how the protocol is run. This is decided by a **Technical Fellowship** 

- The **Technical Fellowship** is limited to working on technical upgrades. The fellowship team is small in size, probably less than 5 members to facilitate consensus. Unlike Polkadot the community doesn't vote on technical upgrades. Also, the TEA Technical Fellowship doesn't have anything to do with speeding referendums through or any other aspect of voting.
- The Fellowship alone decides on technical details to implement and what they work on is not up for vote. What is up for vote is the actual Fellowship Team.
- The community can impeach the TEA fellowship through voting if they're dissatisfied with their performance. For the first 2 years after mainnet launch there will be no vote, as it will be the TEA core technical team founders who implement the vote. After that, the community can vote to impeach the fellowship team and elect a new fellowship team.

**Comment by geopor** [comment]
forum discussion = vetting
snapshot.org = more formal off-chain governance


DAO should have both. Discussion stage is for proposing and vetting ideas of community members and deciding which ideas are ready to go to “phase 2: Vote on Snapshot (off-chain)” for more formal voting.

**Comment by geopor** [comment]
Also, to be fully decentralized (no council members with seniority above other members) we could follow how Polkadot does it: https://polkadot.network/blog/gov2-polkadots-next-generation-of-decentralised-governance/

**Comment by geopor** [comment]
.

**Reply by kevingzhang** [reply]
if the reputation token is in TEA layer2, we will not be able to use snapshot.org to vote. Because the snapshot can use ERC20 token layer1 only. I am not sure if it is possible to have an API to use our layer2 token.

if we allow user to withdraw reputation token to layer1, this will be some game the system by selling the reputation. so we may have to disable any topup/withdraw/transfer of reputation token. it has to be a soul-bonding token

@geopor do you have a solution?

**Reply by geopor** [reply]
Users could mint a **reputation score** to their layer1 address based on their layer2 reputation. This would be a soulbound token that stores the user's reputation score through the metadata. A mint on layer2 adds the score to layer1 while removing it from layer2.

**Comment by kevingzhang** [comment]
Please update the original design @geopor

**Comment by geopor** [comment]
Updated

**Comment by kevingzhang** [comment]
we can use this formula to calculate the reputation token gain in a day.
```
( invested_tea_in_all_bonding_curve / 365 
 + node_owner_paid_state_sub_tax
iu + state_maintainer_paid_harberger_tax
 + app_owner_paid_memory_tax )
* square_root(your_reputation_balance)/your_reputation_balance
```

this can make the the higher your reputation balance, the slower your gain new reputation. 



I would like to add another rule: If you have not vote/propose any DAO voting events for X months. Your reputation won't continue grow. 


@geopor if you agree, please update

**Reply by geopor** [reply]
Updated
